mava/networks/utils/sable/encoder_decoder_fns.py

```diff
@@ -0,0 +1,271 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Tuple
+
+import chex
+import distrax
+import jax
+import jax.numpy as jnp
+from flax import linen as nn
+
+# General shapes legend:
+# B: batch size
+# S: sequence length
+# C: number of agents per chunk of sequence
+# N: number of actions
+# A: number of agents
+
+
+def train_encoder_fn(
+    encoder: nn.Module,
+    obs: chex.Array,
+    hstate: chex.Array,
+    dones: chex.Array,
+    step_count: chex.Array,
+    chunk_size: int,
+) -> Tuple[chex.Array, chex.Array, chex.Array]:
+    """Chunkwise encoding for discrete action spaces."""
+    B, S = obs.shape[:2]
+    v_loc = jnp.zeros((B, S, 1))
+    obs_rep = jnp.zeros((B, S, encoder.net_config.embed_dim))
+
+    # Apply the encoder per chunk
+    num_chunks = S // chunk_size
+    for chunk_id in range(0, num_chunks):
+        start_idx = chunk_id * chunk_size
+        end_idx = (chunk_id + 1) * chunk_size
+        chunk_obs = obs[:, start_idx:end_idx]
+        chunk_dones = dones[:, start_idx:end_idx]
+        chunk_step_count = step_count[:, start_idx:end_idx]
+        chunk_v_loc, chunk_obs_rep, hstate = encoder(
+            chunk_obs, hstate, chunk_dones, chunk_step_count
+        )
+        v_loc = v_loc.at[:, start_idx:end_idx].set(chunk_v_loc)
+        obs_rep = obs_rep.at[:, start_idx:end_idx].set(chunk_obs_rep)
+
+    return v_loc, obs_rep, hstate
+
+
+def train_decoder_fn(
+    decoder: nn.Module,
+    obs_rep: chex.Array,
+    action: chex.Array,
+    legal_actions: chex.Array,
+    hstates: chex.Array,
+    dones: chex.Array,
+    step_count: chex.Array,
+    n_agents: int,
+    chunk_size: int,
+    rng_key: Optional[chex.PRNGKey] = None,
+) -> Tuple[chex.Array, chex.Array]:
+    """Parallel action sampling for discrete action spaces."""
+    # Delete `rng_key` since it is not used in discrete action space
+    del rng_key
+
+    shifted_actions = get_shifted_actions(action, legal_actions, n_agents=n_agents)
+
+    logit, _ = act_chunkwise(
+        decoder=decoder,
+        obs_rep=obs_rep,
+        shifted_actions=shifted_actions,
+        hstates=hstates,
+        dones=dones,
+        step_count=step_count,
+        legal_actions=legal_actions,
+        chunk_size=chunk_size,
+    )
+
+    masked_logits = jnp.where(
+        legal_actions,
+        logit,
+        jnp.finfo(jnp.float32).min,
+    )
+
+    distribution = distrax.Categorical(logits=masked_logits)
+    action_log_prob = distribution.log_prob(action)
+    action_log_prob = jnp.expand_dims(action_log_prob, axis=-1)
+    entropy = jnp.expand_dims(distribution.entropy(), axis=-1)
+
+    return action_log_prob, entropy
+
+
+def act_chunkwise(
```

Comment:
Can we keep the names the same as training:
```suggestion
def act_decoder_fn(
```

Reply:
forgot to update it, nice catch!